lib/JsonClient.js

```diff
@@ -57,18 +60,22 @@ JsonClient.prototype.parse = function parse(req, callback) {
     function parseResponse(err, req2, res, data) {
         var obj;
         var resErr = err;
+        var parseErr;
 
         try {
             if (data && !/^\s*$/.test(data)) {
                 obj = JSON.parse(data);
             }
         } catch (e) {
-            // Not really sure what else we can do here, besides
-            // make the client just keep going.
-            log.trace(e, 'Invalid JSON in response');
+            // bad data being returned that cannot be parsed should be surfaced
+            // to the caller. http errors should take precedence, but it's
+            // possible to receive malformed data regardless of a status code.
+            parseErr = e;
+            log.trace(parseErr, 'Invalid JSON in response');
         }
-        obj = obj || {};
+        obj = obj || data;
```

Comment:
I think we are almost there. Now, for an empty response, we get an empty string instead of an empty object but we don't get a parse error.

Reply:
Yeah, that stems from the legacy issue discussed above. I thought about it some, since we're going to breaking behavior anyway we might as well do it now to keep it strict/consistent with what's valid JSON. I'll make that change.

Reply:
Actually, if it's an empty response then there's nothing to parse. I'm not sure this should be a parse error (but I have made whitespace a parse error).

Reply:
Ah, sorry, I may not have been clear on this one. I was actually okay keeping the old behavior (whitespace in the body is ignored) but would like to add an else above on line 68 where we default the empty string to an Object keeping backwards compatibility. The changes in this PR cause the library to return an empty string instead of an empty object when we get an empty response back from the server. I suspect this may wreak havoc somewhere in userland.

`JSON.parse('')` will yield `SyntaxError: Unexpected end of JSON input`, which I'm not sure is what we want.